app/main.py

```diff
@@ -2,43 +2,45 @@ class Distance:
     def __init__(self, km):
         self.km = km
 
+    @staticmethod
+    def check_type(other):
+        if isinstance(other, Distance):
+            return other.km
+        return other
+
     def __str__(self):
         return f'Distance: {self.km} kilometers.'
 
     def __repr__(self):
         return f'Distance(km={self.km})'
 
     def __add__(self, other):
-        if isinstance(other, Distance):
-            km = self.km + other.km
-        else:
-            km = self.km + other
-        return Distance(km)
+        return Distance(self.km + self.check_type(other))
 
     def __iadd__(self, other):
-        self.km = self.__add__(other).km
+        self.km = self.__add__(other)
```

Comment:
Regarding object assignment - you've requested to make a type check earlier. 
So now we have a type check in __add__ method. If Distance object is passed, it will pass Distance.km

Regarding magic methods - understood, thank you.